pkg/util/chiprometheus/middleware.go

```diff
@@ -0,0 +1,62 @@
+package chiprometheus
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-chi/chi/middleware"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	dflBuckets = []float64{300, 1200, 5000}
+)
+
+const (
+	reqsName    = "_requests_total"
+	latencyName = "_request_duration_milliseconds"
+)
+
+type Middleware struct {
+	reqs    *prometheus.CounterVec
+	latency *prometheus.HistogramVec
+}
+
+func NewMiddleware(name string, buckets ...float64) func(next http.Handler) http.Handler {
+	var m Middleware
+	m.reqs = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        strings.Replace(name+reqsName, "-", "_", -1),
+			Help:        "How many HTTP requests processed, partitioned by status code, method and HTTP path.",
+			ConstLabels: prometheus.Labels{"service": name},
+		},
+		[]string{"code", "method", "path"},
+	)
+	prometheus.MustRegister(m.reqs)
+
+	if len(buckets) == 0 {
+		buckets = dflBuckets
+	}
+	m.latency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:        strings.Replace(name+latencyName, "-", "_", -1),
+		Help:        "How long it took to process the request, partitioned by status code, method and HTTP path.",
+		ConstLabels: prometheus.Labels{"service": name},
+		Buckets:     buckets,
+	},
+		[]string{"code", "method", "path"},
+	)
+	prometheus.MustRegister(m.latency)
+	return m.handler
+}
+
+func (c Middleware) handler(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
+		next.ServeHTTP(ww, r)
+		c.reqs.WithLabelValues(http.StatusText(ww.Status()), r.Method, r.URL.Path).Inc()
+		c.latency.WithLabelValues(http.StatusText(ww.Status()), r.Method, r.URL.Path).Observe(float64(time.Since(start).Nanoseconds()) / 1000000)
```

Comment:
Same as above here.

Reply:
Fixed in 7809a812fd399631d25aaa248936ea830a4ff796